File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int start = 0, end = 0, count = 0;
        long product = 1; // Using long to prevent integer overflow
        
        while (end < nums.length) {
            product *= nums[end];
            
            while (start <= end && product >= k) {
                product /= nums[start];
                start++;
            }
            
            if (product < k)
                count += end - start + 1;
            
            end++;
        }
        
        return count;
    }
}

File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        start = 0
        end = 0
        count = 0
        product = 1
    
        while end < len(nums):
            product *= nums[end]
        
            while start <= end and product >= k:
                product /= nums[start]
                start += 1
        
            if product < k:
                count += end - start + 1
        
            end += 1
    
        return count